[WillCWX] iP

WillCWX opened the pull request:
# Upgraded Duke Bot
```
Hello from
 ____        _        
|  _ \ _   _| | _____ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|
```
> This is all duke can do in your master branch.

The upgraded duke bot can
- Echo user input
- Add
  1.   Todo tasks
  2.   Deadline tasks
  3.   Event tasks
- Mark tasks as done
- Unmark tasks and make them not done.
- Delete tasks
- List out all your task
- Save single user sessions
- Load single user sessions
- Priorities tasks to print out
- Find tasks with a name

With this new duke:

- [X] No more typing out everything for todo lists
- [X] More ~steps~ fun with duke  
- [X] [Handsome Duke](https://user-images.githubusercontent.com/55905659/216753701-29ad1368-b55e-4046-8b70-c3507ad1876f.png) 😍  to talk to _(ayo?)_

 Just run the `main` from the `Launcher.java` to run duke now!
 ```Java
 public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
 ```

WillCWX committed:
Added Duke's Greeting, Echo and Exit

WillCWX committed:
Changed command and reply format

WillCWX committed:
Completed #1

WillCWX committed:
Completes and closes #2 with Duke's list and add function

WillCWX committed:
Added and implemented Task class to Duke

WillCWX committed:
Closes #3, refactored interpret command

WillCWX committed:
Added Deadline, Event and ToDo classes

WillCWX committed:
Implemented ToDo, Deadline, Event into Duke

WillCWX committed:
Implemented Todo, Deadline, Event into Duke

WillCWX committed:
Closes #4 and cleaned up command scanning

WillCWX committed:
Resolves #9, reformats code and documentation

WillCWX committed:
Closes #7, updates input and EXPECTED

WillCWX committed:
Closes #5, creates and implements DukeException

Errors in user input are not caught and runtime exceptions are thrown.

This not only causes Duke to exit ungracefully but also causes the
unsmooth operation of duke.

Thus, I create a DukeException class, extending the Exception
class, which can be thrown with an error message.

Having our own checked exception allows our own exception to be thrown,
caught and then handled to tell the user what was wrong.

This is also done to conform to Level-5
https://nus-cs2103-ay2223s2.github.io/website/schedule/week2/project.html

WillCWX committed:
Closes #6, creates deleteTask method

Users cannot delete any task after they are given to Duke.

This is an issue to the user as the task list becomes too big and
cumbersome to read. Furthermore, any erronous tasks added cannot be
changed or deleted.

Hence, I replaced the Task[] array with an ArrayList<Task> collection.
Then, I created the deleteTask method that deletes a task using the
ArrayList methods.

This will resolve the current issue and also keep the ordering of the
tasks. Using collections also makes deleting withou errors much easier.

WillCWX committed:
Solves #8 by creating and implementing Enum

WillCWX committed:
Resolves test failure in linux

While the test passed in Windows, it failed in linux due to a missing \n character.
Also, this was done in vim on linux, on a phone, as a test.

WillCWX committed:
Implements serialization to task, storage class to handle files

There is no way to preserve sessions. Users have to retype everything
which is a pain.

I create a storage class to deal with files, both storing and loading.
Writing to a file and reading from a file becomes easier because of
serialization.

This should make saving and loading possible and easier to implement.

WillCWX committed:
Implements save and load in Duke

The Storage class has been created but was not added to Duke yet.

This commit adds the Storage class functions to Duke, allowing Duke to
now save and load. This preserves user sessions nicely.

WillCWX committed:
Adds a DatedTask for comparable implementation

WillCWX committed:
Replaces string dates with localdate

WillCWX committed:
Updates EXPECTED.TXT to new LocalDate

WillCWX committed:
Merge branch 'branch-Level-8'

WillCWX committed:
Separates Ui from Duke

WillCWX committed:
Refactors Duke functions into classes

Huge changes

WillCWX committed:
Moves classes into respective packages

WillCWX committed:
Adds gradle project manager

WillCWX committed:
Adds JUnit tests for some methods

WillCWX committed:
Adds and fixes JavaDocs

WillCWX committed:
Adds the Find Command

WillCWX committed:
Merge branch 'branch-Level-9'

Creationsv2 commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,32 @@
+    /** Constructs the add deadline command. */
+    public AddDeadlineCommand() {}
+
+    /**
+     * Adds a Deadline task to the tasklist and stores the new tasklist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
Perhaps you would like to add `@param` descriptors in the JavaDocs

Creationsv2 commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,30 @@
+    /** Constructs the add event command. */
+    public AddEventCommand() {}
+
+    /**
+     * Adds an Event task to the tasklist and stores the new tasklist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
Some '@param' descriptors as per the coding standard

Creationsv2 commented on src/main/java/duke/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,30 @@
+public class AddToDoCommand extends Command {
+    public AddToDoCommand() {}
+
+    /**
+     * Adds a ToDo task to the tasklist and stores the new tasklist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
'@param' descriptors as per the coding standard

Creationsv2 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return False.
+     */
+    public boolean isExit() {
```
Perhaps a verb would be a better method name, maybe `exitStatus()`?

WillCWX replied:
Thanks for the suggestion, name changed to `canExit`

Creationsv2 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,28 @@
+    /** Constructs the delete command. */
+    public DeleteCommand() {}
+
+    /**
+     * Deletes a task and stores the resulting tasklist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
Perhaps some `@params` would be more explicit.

Creationsv2 commented on src/main/java/duke/command/EchoCommand.java:
```diff
@@ -0,0 +1,23 @@
+    /** Constructs the echo command. */
+    public EchoCommand() {}
+
+    /**
+     * Echoes the user input.
+     *
+     */
```
Some `@param`s as per coding standard

Creationsv2 commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,34 @@
+    /** Constructs the exit command. */
+    public ExitCommand() {}
+
+    /**
+     * Begins Duke's shutdown.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
`@param`s as per coding standard

Creationsv2 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,27 @@
+    /** Constructs the find command, */
+    public FindCommand() {}
+
+    /**
+     * Prints out the tasks that contains a specific string given by the user.
+     *
+     * @throws DukeException If the string given is empty.
+     */
```
Some `@param`s perhaps

Creationsv2 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,24 @@
+    /** Constructs the list command. */
+    public ListCommand() {}
+
+    /**
+     * Informs the user of the tasks in tasklist.
+     *
+     */
```
Some `@param`s

Creationsv2 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,28 @@
+    /** Constucts the mark command. */
+    public MarkCommand() {}
+
+    /**
+     * Marks the specific task and stores the resulting takslist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
Some `@param`s

Creationsv2 commented on src/main/java/duke/command/PriorityCommand.java:
```diff
@@ -0,0 +1,22 @@
+    /** Constructs the priority command. */
+    public PriorityCommand() {}
+
+    /**
+     * Informs the user of their tasks in order from earliest date not done.
+     *
+     */
```
Some `@param`s

Creationsv2 commented on src/main/java/duke/command/UnknownCommand.java:
```diff
@@ -0,0 +1,26 @@
+    /** Constructs an unknown command. */
+    public UnknownCommand() {}
+
+    /**
+     * Throws a DukeException indicating that the command is not understood.
+     *
+     * @throws DukeException Always as command is not understood.
+     */
```
Some `@param`s

Creationsv2 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,28 @@
+    /** Contructs an unmark command. */
+    public UnmarkCommand() {}
+
+    /**
+     * Unmarks a task and stores the resulting tasklist.
+     *
+     * @throws DukeException If user input is invalid.
+     */
```
Some `@param`s

Creationsv2 commented on src/main/java/duke/task/DatedTask.java:
```diff
@@ -0,0 +1,57 @@
+            throw new DukeException("Write the date in this format: YYYY-MM-DD");
+        }
+    }
+
+    @Override
```
May want to include JavaDoc for this method

Creationsv2 commented on src/main/java/duke/task/DatedTask.java:
```diff
@@ -0,0 +1,57 @@
+    public boolean hasDate() {
+        return true;
+    }
+
+    @Override
```
May want to include JavaDoc for this method

Creationsv2 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,54 @@
+ * The Event class extends the DatedTask class and represents an event with a start and end date.
+ */
+public class Event extends DatedTask {
+    private static final long serialVersionUID = 103;
```
Since this is a constant, may want to name it all uppercase using underscore to separate words e.g. `SERIAL_UID`

Creationsv2 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,54 @@
+                    + "\tevent {name} /from {YYYY-MM-DD} /to {YYYY-MM-DD}");
+        }
+    }
+
+    @Override
```
May want to include JavaDoc for this method

Creationsv2 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,94 @@
+ * This Task class represents a task.
+ */
+public class Task implements Serializable, Comparable<Task> {
+    private static final long serialVersionUID = 100;
```
Since this is a constant, may want to name it all uppercase using underscore to separate words e.g. `SERIAL_UID`

WillCWX replied:
`serialVersionUID` is a special variable found in the Serializable Interface that helps to determine the Class version. Thus, the name cannot change without interfering with the function of the Class.

Creationsv2 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,94 @@
+    public boolean hasDate() {
+        return false;
+    }
+
+    @Override
```
May want to include JavaDoc for this method

Creationsv2 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,121 @@
+ * The tasklist represents the list of tasks.
+ */
+public class TaskList extends ArrayList<Task> {
+    private static final long serialVersionUID = 200;
```
Since this is a constant, may want to name it all uppercase using underscore to separate words e.g. `SERIAL_UID`

WillCWX replied:
>`serialVersionUID` is a special variable found in the Serializable Interface that helps to determine the Class version. Thus, the name cannot change without interfering with the function of the Class.

As per above.

Creationsv2 commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,23 @@
+    public ToDo(String name) {
+        super(name);
+    }
+
+    @Override
```
May want to include JavaDoc for this method

Creationsv2 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,183 @@
+    private Ui() {
+        sc = new Scanner(System.in);
+    }
+
+    public static Ui getInstance() {
```
May want to include JavaDoc for this method

Creationsv2 reviewed (commented):
I was assigned to give comments on coding standard related issues only, but I liked how you went the extra mile to separate out all the commands 👍 

I'm not too sure on the exact syntax for JavaDocs in overridden methods, thus I kept highlighting the `@param` parts for some. If what you did was in actual fact correct, apologies on my part!

WillCWX commented:
Hello @Creationsv2 and @MrTwit99, 
Thanks for the review. 

![image](https://user-images.githubusercontent.com/55905659/216088565-3e32e867-52be-484d-af0b-5640e3c03f92.png)
> [The code issue link](https://github.com/nus-cs2103-AY2223S2/ip/pull/49#discussion_r1090420297)

I have generated some javadocs using gradle and as you can see, the overriding method of the subclass inherits the superclass documentation for the method, even without explicit calls using `{@inheritDoc}`. It shows up normally as well in my IDE. 

However I do understand the issue here and ~I will bring this up in the forum for discussion and judgement later~ I have brought this up in the forum.
https://github.com/nus-cs2103-AY2223S2/forum/issues/124#issue

Thank you

(Edited to tag second reviewer, link to forum discussion)

MrTwit99 commented on src/main/java/duke/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,30 @@
+        store.saveToFile(tasks);
+        ui.printWithPartition("\tGot it. I've added this task:\n" + "\t  " + temp.toString()
+                + "\n\tNow you have " + Integer.toString(tasks.size()) + " tasks in the list.\n");
+
```
Perhaps there is an unnecessary space here that could be removed? I'm not sure if the line is meant for anything as it seems to be empty line on my end
```suggestion
```

MrTwit99 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage store) throws DukeException {
+        String s = ui.getName();
```
Perhaps you could provide another name for variable s such as subString to make it easier to understand.

WillCWX replied:
Thanks, I will follow your suggestion.

MrTwit99 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,28 @@
+package duke.command;
+
+import duke.exception.DukeException;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
```
I'm not sure if it's needed but perhaps you could leave a line gap after each import line if the next import is from different directory. IIRC 2030S used to do it, not sure if needed in 2103T.
```suggestion
import duke.exception.DukeException;

import duke.storage.Storage;

import duke.task.TaskList;

import duke.ui.Ui;
```

WillCWX replied:
![image](https://user-images.githubusercontent.com/55905659/216230005-c55c497d-356e-438f-b9cf-683edcf800f2.png)
According to [se-education tips for IntelliJ IDEA](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order) it appears that the blank lines are only for between `static imports`, `java`, `javax`, `org` and `com` imports.

Though unfortunately the SE-Education Standard for Java does not explicitly write as such :/

MrTwit99 reviewed (commented):
Overall, great job with your iP and I like how you have the commands separated into respective ones.

As for suggestions for improvement, do take a look at some of the ones I listed and I share similar thoughts as the other reviewer on including @param for some of your methods.

WillCWX committed:
Closes #11, closes #14, updates style and JavaDoc

WillCWX committed:
Cleans javadocs and style further

WillCWX committed:
Adds gui skeleton using javafx

WillCWX committed:
Adds skeletal dialogbox to Duke

WillCWX committed:
Adds iteration 1, echoing user

WillCWX committed:
Implements iteration 3, pictures and user input

WillCWX committed:
Moves Duke dialogs left

WillCWX committed:
Adds JavaFXML for easier styling

WillCWX committed:
Implements Duke functions in GUI

WillCWX committed:
Closes #15 by fully implementing JavaFX

WillCWX committed:
Closes #16, implements trival vaargs in find

WillCWX committed:
Updates Gradle mainClassName

WillCWX committed:
Adds assertions to Duke

Currently, no assertions are used in Duke's code.
Assertions are useful a useful lightweight tool that ensures that
assumptions are kept and not violated.

Let's add some assertions to Duke's code. Particularly in places that
assume values are not null or not empty.

This should make Duke less error prone and ensures assumptions are
always kept no matter the version of Duke.

WillCWX committed:
Refactors DatedTasks to improve code quality

The compareTo function in DatedTasks has multiple nested if else that
determine the logic of how DatedTasks are sorted.

This is bad as this makes the logic difficult to follow and understand.

Let's bring out the nested ifs and remove the else by relying on the
return statement to stop the flow of instruction.

This makes the logic of the compareTo function easier to follow and
imporves the code quality.

WillCWX committed:
Adds JavaDocs to DialogBox

There is no JavaDocs for the public methods of getUserDialog and
getDukeDialog.

Public methods should always have JavaDocs unless they are getter/
setters or have inherited JavaDocs.

Let's add the JavaDocs to these two methods.

This will ensure that the behaviour of the method is understood by the
user of the method.

WillCWX committed:
Refactors toCompare method in Task.java

The toCompare method has multiple nested if else that determine the
logic when Tasks are sorted.

These multiple nested if else makes the logic difficult to understand
and the consequent behavour difficult to predict.

Let's bring out the ifs and remove the else.

This makes ensures that the logic is better understood and the behavour
is more easily predicted.

The elses can be removed as we can use the return statements to stop the
instruction flow.

WillCWX committed:
Merge pull request #18 from WillCWX/branch-A-Assertions

Adds assertions to Duke

WillCWX committed:
Merge pull request #19 from WillCWX/branch-A-CodeQuality

Branch a code quality

WillCWX committed:
Adds CI to Duke

Currently, Duke is not continuously integrated for any updates pushed or
merged in pull requests.

This means that it is possible for a version of Duke to not work and
have devlopers/users not realise.

Let's set up a CI for Duke.

This will automatically make sure that Duke is compilable in gradle and
therefore inform developers/users if the latest changes made have broken
Duke.

WillCWX committed:
Makes gradlew executable

WillCWX committed:
Adds Undo function to Duke

Users are unable to undo their previous commands.

If users made a mistake, they have to remove what they did previously
using the opposite command. This is troublesome.

Let's add an undo function to duke.

This will allow users to correct their mistakes easier.

WillCWX committed:
Removes deprecated Ui class from Duke

Duke currently retains the Ui class from its previous functions as a
CLI chat bot.

However, as new updates are given to the GUI form of Duke, the CLI form
is no longer maintained as well. The CLI form now lacks the Undo function
that is already implemented for the GUI.

Let's delete deprecated functions that still use Ui.

This will reduce the errors and is part of the deprecation process of Ui.

WillCWX committed:
Removes Ui class

WillCWX committed:
Fixes Undo function

WillCWX committed:
Adds User Guide
